src/data-model-providers/codegen/ServerClusterInterfaceRegistry.cpp

```diff
@@ -0,0 +1,224 @@
+            {
+                mCachedClusterEndpointId = kInvalidEndpointId;
+                mCachedInterface         = nullptr;
+            }
+
+            ServerClusterInterface * result = current->serverClusterInterface;
+            Platform::MemoryFree(current);
+
+            return result;
+        }
+
+        prev    = current;
+        current = current->next;
+    }
+
+    // Not found.
+    return nullptr;
+}
+
+ServerClusterInterfaceRegistry::ClustersList ServerClusterInterfaceRegistry::ClustersOnEndpoint(EndpointId endpointId)
+{
+    EndpointClusters * clusters = FindClusters(endpointId);
+
+    if (clusters == nullptr)
+    {
+        return nullptr;
+    }
+
+    return clusters->firstCluster;
+}
+
+void ServerClusterInterfaceRegistry::UnregisterAllFromEndpoint(EndpointId endpointId)
+{
+    if (mCachedClusterEndpointId == endpointId)
+    {
+        // all clusters on the given endpoint will be unregistered.
+        mCachedInterface         = nullptr;
+        mCachedClusterEndpointId = kInvalidEndpointId;
+    }
+
+    EndpointClusters * prev    = nullptr;
+    EndpointClusters * current = mEndpoints;
+    while (current != nullptr)
+    {
+        if (current->endpointId == endpointId)
+        {
+            if (prev == nullptr)
+            {
+                mEndpoints = current->next;
+            }
+            else
+            {
+                prev->next = current->next;
+            }
+            ClearSingleLinkedList(current->firstCluster);
+            Platform::Delete(current);
+            return;
+        }
+
+        prev    = current;
+        current = current->next;
+    }
+}
+
+ServerClusterInterface * ServerClusterInterfaceRegistry::Get(const ConcreteClusterPath & path)
+{
+    EndpointClusters * endpointClusters = FindClusters(path.mEndpointId);
+    VerifyOrReturnValue(endpointClusters != nullptr, nullptr);
+    VerifyOrReturnValue(endpointClusters->firstCluster != nullptr, nullptr);
+
+    // Check the cache to speed things up
+    if ((mCachedClusterEndpointId == path.mEndpointId) && (mCachedInterface != nullptr) &&
+        (mCachedInterface->GetClusterId() == path.mClusterId))
+    {
+        return mCachedInterface;
+    }
+
+    // The cluster searched for is not cached, do a linear search for it
+    RegisteredServerClusterInterface * current = endpointClusters->firstCluster;
+
+    while (current != nullptr)
+    {
+        if (current->serverClusterInterface->GetClusterId() == path.mClusterId)
+        {
+            mCachedClusterEndpointId = path.mEndpointId;
+            mCachedInterface         = current->serverClusterInterface;
+            return mCachedInterface;
+        }
+
+        current = current->next;
+    }
+
+    // not found
+    return nullptr;
+}
+
+ServerClusterInterfaceRegistry::EndpointClusters * ServerClusterInterfaceRegistry::FindClusters(EndpointId endpointId)
+{
+    // invalid cluster id is NOT acceptable (since static allocation uses the
```

Comment:
nit, I think this comment and should be saying invalid endpoint id

Reply:
Removed the whole check: there is no more static allocation anymore here. Hoping this saves a minute amount of flash.